lib/externalGraderCommon.js

```diff
@@ -117,11 +117,14 @@ module.exports.buildDirectory = function(dir, submission, variant, question, cou
                     // the directory hierarchy
                     const fullPath = path.join(dir, 'student', file.name);
                     const relPath = path.relative(path.join(dir, 'student'), fullPath);
-                    if (relPath.split(path.sep)[0] == '..' || path.isAbsolute(relPath)) {
+                    if (relPath.split(path.sep).includes('..') || path.isAbsolute(relPath)) {
                         callback(new Error('Invalid filename'));
                         return;
                     }
-                    fs.writeFile(path.join(dir, 'student', file.name), decodedContents, callback);
+                    // We need to make sure that the directories are created first.
+                    // Here fs is actually fs-extra, so we can use ensureDirSync.
+                    fs.ensureDirSync(path.dirname(fullPath));
+                    fs.writeFile(fullPath, decodedContents, callback);
```

Comment:
Can we convert the `ensureDirSync` call to callback style?  I recall that sync code is avoided in the server as it can lead to performance issues.

Reply:
@nicknytko How about now? I put an async/await wrapper.

Reply:
Hold on, this isn't valid. I'll fix it.

Reply:
I think it's correct now. I removed the wrapper (I hadn't properly included the file write within the async function to begin with). Now it's just using the Promise's `catch` and `then`.